### PR TITLE
[NRLF 242] inbound outbound errors

### PR DESCRIPTION
**Pull request by Rohoolio**

### CODE REVIEW

File: feature_tests/features/consumer/consumer-readDocumentPointer-edge.feature [hunk 1]
```diff
@@ -87,9 +87,9 @@ Feature: Consumer Read Edge Case scenarios
     When Consumer "Yorkshire Ambulance Service" reads an existing Document Reference "8FW23-1114567891"
     Then the operation is unsuccessful
```
Comment (deepaattiguppe) [inline feature_tests/features/consumer/consumer-readDocumentPointer-edge.feature:88]: Can we please add the step to test the status code to all the tests in the PR where it is missing - **And the status is '400'.**

Comment (Rohoolio) [reply feature_tests/features/consumer/consumer-readDocumentPointer-edge.feature:88]: Yeah can do :)

File: layer/nrlf/nrlf/core/authoriser.py [hunk 2]
```diff
@@ -74,11 +73,25 @@ def parse_headers(
     _raw_connection_metadata = _headers.get(CONNECTION_METADATA, "{}")
     try:
         connection_metadata = ConnectionMetadata.parse_raw(_raw_connection_metadata)
+    except ValidationError:
+        return PipelineData(
+            error={
```
Comment (jaklinger) [inline layer/nrlf/nrlf/core/authoriser.py:78]: seems like we never need the intermediate dict `message`, why not just `"error": "blah"`

File: layer/nrlf/nrlf/core/authoriser.py [hunk 3]
```diff
@@ -74,11 +73,25 @@ def parse_headers(
...
+                "message": "There was an issue parsing the connection metadata, contact onboarding team"
```
Comment (jaklinger) [inline layer/nrlf/nrlf/core/authoriser.py:79]: I'd set up a const to share between the two:

```
PARSE_ERROR_MESSAGE = "There was an issue parsing the {model_name}, contact onboarding team"
```

Which you can fill with:

```
PARSE_ERROR_MESSAGE.format(model_name="connection metadata")
```

File: layer/nrlf/nrlf/core/validators.py [hunk 4]
```diff
@@ -82,9 +83,13 @@ def validate_document_reference_string(fhir_json: str):
...
     except ValidationError:
-        raise DocumentReferenceValidationError("Item could not be found")
+        raise DocumentReferenceValidationError(
+            "There was a problem retrieving the document reference"
+        )
     except ValueError:
-        raise DocumentReferenceValidationError("Item could not be found")
+        raise DocumentReferenceValidationError(
+            "There was a problem retrieving the document reference"
+        )
```
Comment (jaklinger) [inline layer/nrlf/nrlf/core/validators.py:92]: Two recommendations in one:

```
    except (ValidationError, ValueError):
        raise DocumentReferenceValidationError(
            "There was a problem retrieving the document reference"
        ) from None
```

File: api/producer/createDocumentReference/src/v1/handler.py [hunk 5]
```diff
@@ -109,15 +109,15 @@ def validate_producer_permissions(
         core_model=core_model,
         pointer_types=data["pointer_types"],
     ):
-        raise AuthenticationError(
-            "Required permissions to create a document pointer are missing"
+        raise RequestValidationError(
+            "Required permissions to create a document pointer are missing or mismatched"
         )
```
[the remark this answers is not in the document]
Comment (Rohoolio) [reply api/producer/createDocumentReference/src/v1/handler.py:114]: That message isnt the only one associated with the RequestValidationError - we have a couple that are based around the relatesTo fields as well, unless you suggest we make a different error for that?

File: api/producer/updateDocumentReference/src/v1/handler.py [hunk 6]
```diff
@@ -37,7 +37,9 @@ def parse_request_body(
     body = fetch_body_from_event(event)
 
     if ("id" in body and "id" in data) and (body["id"] != data["id"]):
-        raise ItemNotFound("Item could not be found")
+        raise RequestValidationError(
+            "Existing document id does not match the document id in the body"
```
[the remark this answers is not in the document]
Comment (Rohoolio) [reply api/producer/updateDocumentReference/src/v1/handler.py:41]: I did consider that some of these errors are slightly different, but i dont think this ones different enough to warrant a whole new error mapping - Id rather adjust the wording if that was the issue? the conditional itself was a little confusing so i just wrote what i thought went wrong there haha

**Review by brianbayes — APPROVED** [review]

File: layer/nrlf/nrlf/core/common_producer_steps.py [hunk 7]
```diff
@@ -21,7 +22,14 @@ def validate_producer_permissions(
...
+            "The target document reference does not belong to the requesting organisation"
         )
     return PipelineData(**data)
+
+
+def invalid_producer_for_delete(organisation_code, delete_item_id: str):
+    producer_id = generate_producer_id(id=delete_item_id, producer_id=None)
+    if not organisation_code == producer_id:
+        return True
+    return False
```
Comment (jaklinger) [inline layer/nrlf/nrlf/core/common_producer_steps.py:35]:
```suggestion
    return organisation_code == producer_id
```

File: layer/nrlf/nrlf/core/tests/test_transform.py [hunk 8]
```diff
@@ -62,6 +64,41 @@ def test__strip_empty_json_paths_do_not_raise(input):
...
+    ),
+)
+def test_strip_empty_json_paths_throws_error_when_field_missing(
+    field, expected_outcome
+):
+    fhir_json = read_test_data("nrlf")
+    fhir_json[field] = None
+
+    with pytest.raises(expected_outcome):
+        _strip_empty_json_paths(fhir_json, raise_on_discovery=True)
```
Comment (jaklinger) [inline layer/nrlf/nrlf/core/tests/test_transform.py:99]: can reduce to

```suggestion
@pytest.mark.parametrize("field", REQUIRED_CREATE_FIELDS)
def test_strip_empty_json_paths_throws_error_when_field_missing(field):
    fhir_json = read_test_data("nrlf")
    fhir_json[field] = None

    with pytest.raises(ProducerCreateValidationError):
        _strip_empty_json_paths(fhir_json, raise_on_discovery=True)
```

maybe plus one test that `REQUIRED_CREATE_FIELDS == ["custodian", "id", "type", "status", "subject"]`

File: layer/nrlf/nrlf/core/authoriser.py [hunk 9]
```diff
@@ -74,11 +73,21 @@ def parse_headers(
...
+    try:
         ClientRpDetailsHeader.parse_raw(_raw_client_rp_details)
-    except ValidationError as err:
-        return PipelineData(error={"message": get_error_message(err)}, **data)
-    else:
-        return PipelineData(pointer_types=connection_metadata.pointer_types, **data)
+    except ValidationError:
+        return PipelineData(
+            error="There was an issue parsing the client rp details, contact onboarding team"
+            ** data,
```
[the remark this answers is not in the document]
Comment (Rohoolio) [reply layer/nrlf/nrlf/core/authoriser.py:87]: How has that not caused some weird ass behaviour?

**Pull request merged**